File: src/test/java/com/running/dto/buddy/test/java/cases/validate/PartiallyImplementedPropertyValidationTest.java
package com.running.dto.buddy.test.java.cases.validate;

import com.runninglane.dto.buddy.DtoBuddy;
import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertThrows;

public class PartiallyImplementedPropertyValidationTest {

    abstract class AbstractClassWithConcreteGetterAndAbstractSetter {
        public String getName() {
            return "John";
        }

        public abstract void setName(String name);
    }

    abstract static
    class AbstractClassWithAbstractGetterAndConcreteSetter {
        public abstract String getName();

        public void setName(String name) {
            throw new UnsupportedOperationException("Not yet implemented");
        }
    }

    @Test
    public void shouldFailWhenGetterIsConcreteAndSetterIsAbstract() {
        assertThrows(DtoBuddyBadInputException.class, () ->
            DtoBuddy.implementor(AbstractClassWithConcreteGetterAndAbstractSetter.class).implement()
        );
    }

    @Test
    public void shouldFailWhenGetterIsAbstractAndSetterIsConcrete() {
        assertThrows(DtoBuddyBadInputException.class, () ->
            DtoBuddy.implementor(AbstractClassWithAbstractGetterAndConcreteSetter.class).implement()
        );
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/validate/PartiallyImplementedPropertyValidationTest.kt
package com.runninglane.dto.buddy.test.cases.validate

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.exception.DtoBuddyBadInputException
import org.junit.jupiter.api.assertThrows
import kotlin.test.Test

class PartiallyImplementedPropertyValidationTest {

    abstract class AbstractClassWithConcreteGetterAndAbstractSetter {
        fun getName(): String = "John"
        abstract fun setName(name: String)
    }

    abstract class AbstractClassWithAbstractGetterAndConcreteSetter {
        abstract fun getName(): String
        fun setName(name: String) { TODO("Not yet implemented") }
    }

    @Test
    fun shouldFailWhenGetterIsConcreteAndSetterIsAbstract() {
        assertThrows<DtoBuddyBadInputException> {
            DtoBuddy.implement(AbstractClassWithConcreteGetterAndAbstractSetter::class.java)
        }
    }

    @Test
    fun shouldFailWhenGetterIsAbstractAndSetterIsConcrete() {
        assertThrows<DtoBuddyBadInputException> {
            DtoBuddy.implement(AbstractClassWithAbstractGetterAndConcreteSetter::class.java)
        }
    }
}